lib/src/native/realm_core.dart

```diff
@@ -450,6 +453,53 @@ class _RealmCore {
     return RealmHandle._(realmPtr);
   }
 
+  RealmSchema readSchema(Realm realm) {
+    return using((Arena arena) {
+      final classKeys = _getValues<Uint32, int>(
+          arena,
+          (size) => arena<Uint32>(size),
+          (valuesPtr, size, outSize) => _realmLib.realm_get_class_keys(realm.handle._pointer, valuesPtr, size, outSize),
+          (values, index) => values.elementAt(index).value);
+
+      final result = classKeys.map((e) => _getSchemaForClassKey(realm, e, arena)).toList();
+      return RealmSchema(result);
+    });
+  }
+
+  SchemaObject _getSchemaForClassKey(Realm realm, int classKey, Arena arena) {
+    final classInfo = arena<realm_class_info>();
+    _realmLib.invokeGetBool(() => _realmLib.realm_get_class(realm.handle._pointer, classKey, classInfo));
+
+    final name = classInfo.ref.name.cast<Utf8>().toDartString();
+
+    final nativeProperties = _getValues<realm_property_info, realm_property_info>(
+        arena,
+        (size) => arena<realm_property_info>(size),
+        (valuesPtr, size, outSize) => _realmLib.realm_get_class_properties(realm.handle._pointer, classKey, valuesPtr, size, outSize),
+        (values, index) => values.elementAt(index).ref);
+
+    final properties = nativeProperties.map((e) => e.toSchemaProperty()).toList();
+
+    return SchemaObject(RealmObject, name, properties);
+  }
+
+  List<TReturn> _getValues<T extends NativeType, TReturn>(Arena arena, Pointer<T> Function(int size) createPtr,
+      bool Function(Pointer<T> valuesPtr, int size, Pointer<IntPtr> outSize) getValue, TReturn Function(Pointer<T> values, int index) getAtIndex) {
+    // TODO: this is a hack to get the size of the array and should not be needed once https://github.com/realm/realm-core/issues/5430 is addressed.
```

Comment:
This core issue seems to be handled already. https://github.com/realm/realm-core/issues/5430